### PR TITLE
MP groups and permissions

### PR DESCRIPTION
**Pull request by zsilencer**
mp

### CODE REVIEW

**Comment by IntelOrca** [comment]
Impressive.
Currently anyone with full access can set other clients including themselves to spectator / full access. This should probably only be allowed by host. Alternatively there could just an additional "moderator" permission level.

**Comment by zsilencer** [comment]
I plan on expanding the group management tab so that you can create new groups with fine grained control, such as allowing only certain game commands, and boundaries on the map where things can be placed, etc.  Right now just wanted these two defaults, and I made it so the host always has full access, with a setting of a default group of any players joining.  So you can just have you and your buddies working on a park without needing a password, and can let anyone join, watch and chat.

Making everything in the interfaces read-only might be a big pain, and they get a error box saying permission denied when they try things, so it shouldn't be too hard for users to understand.

**Comment by IntelOrca** [comment]
Currently clients with full access can change the permissions of other clients, this shouldn't be allowed for the time being.

**Comment by janisozaur** [comment]
Cool stuff. Out of curiosity – does OpenTTD has something like that?

**Comment by zsilencer** [comment]
@IntelOrca K I have this mostly complete now, and have added a default User group that does not have permission to modify or assign groups to address your concern.  Maybe some people can test this.

**Comment by IntelOrca** [comment]
@LRFLEW could you xcode this PR?

File: src/network/network.cpp [hunk 1]
```diff
@@ -1620,6 +1898,243 @@ int network_get_player_id(unsigned int index)
...
+					}
+				} else {
+					group->ToggleActionPermission(index);
+				}
+			}
+		}
+	}break;
+	case 3:{ // set group name
+		size_t nameChunkOffset = nameChunkIndex - 1;
+		if (nameChunkOffset < 0)
```
Comment (janisozaur) [inline src/network/network.cpp]: `size_t` is unsigned, so this is always false.

File: src/windows/multiplayer.c [hunk 2]
```diff
@@ -0,0 +1,701 @@
...
+
+		if (y + 11 >= dpi->y) {
+			char buffer[300];
+			char* lineCh;
+			int groupindex = network_get_group_index(_selectedGroup);
+			if (groupindex != -1){
+				if (network_can_perform_action(groupindex, i)) {
+					lineCh = buffer;
+					lineCh = utf8_write_codepoint(lineCh, FORMAT_WINDOW_COLOUR_2);
+					safe_strcpy(lineCh, "x", sizeof(buffer) - (lineCh - buffer));
```
Comment (IntelOrca) [inline src/windows/multiplayer.c]: You could use FORMAT_TICK instead here.

File: src/game.c [hunk 3]
```diff
@@ -533,6 +540,10 @@ int game_do_command_p(int command, int *eax, int *ebx, int *ecx, int *edx, int *
 				}
 			}
 
+			if (network_get_mode() == NETWORK_MODE_SERVER && !(flags & GAME_COMMAND_FLAG_NETWORKED)) {
+				network_add_player_money_spent(network_get_current_player_id(), cost);
```
Comment (IntelOrca) [inline src/game.c]: We need to block this for ghosts.

**Comment by LRFLEW** [comment]
@zsilencer I made a PR against your branch to fix the Xcode project. I also fixed one of @IntelOrca's comments, as it was causing a warning in Xcode.

**Comment by zsilencer** [comment]
still have some things to change so do not merge yet

**Comment by IntelOrca** [comment]
This reminds me of Locomotion which showed you what other players were looking at.

**Comment by IntelOrca** [comment]
Clients can still pick up and move guests even though the guest permission is disabled.

**Comment by zsilencer** [comment]
alright i fixed that, let me know if anyone finds any other issues.

**Comment by IntelOrca** [comment]
Does it save the groups and permissions?

**Comment by janisozaur** [comment]
wOOt! Someone post this to /r/openrct2

**Pull request merged**